HW2/P5/physics.pyx

```diff
@@ -0,0 +1,156 @@
+    cdef:
+        float deltax = x2[0] - x1[0]
+        float deltay = x2[1] - x1[1]
+        float vrelx = v2[0] - v1[0]
+        float vrely = v2[1] - v1[1]
+    # check if delta and velocity in same direction
+    return (deltax * vrelx + deltay * vrely) > 0.0
+
+
+cdef inline void collide(FLOAT *x1, FLOAT *v1,
+                         FLOAT *x2, FLOAT *v2) nogil:
+    cdef:
+        float x1_minus_x2[2]
+        float v1_minus_v2[2]
+        float change_v1[2]
+        float len_x1_m_x2, dot_v_x
+        int dim
+    # https://en.wikipedia.org/wiki/Elastic_collision#Two-dimensional_collision_with_two_moving_objects
+    for dim in range(2):
+        x1_minus_x2[dim] = x1[dim] - x2[dim]
+        v1_minus_v2[dim] = v1[dim] - v2[dim]
+    len_x1_m_x2 = x1_minus_x2[0] * x1_minus_x2[0] + x1_minus_x2[1] * x1_minus_x2[1]
+    dot_v_x = v1_minus_v2[0] * x1_minus_x2[0] + v1_minus_v2[1] * x1_minus_x2[1]
+    for dim in range(2):
+        change_v1[dim] = (dot_v_x / len_x1_m_x2) * x1_minus_x2[dim]
+    for dim in range(2):
+        v1[dim] -= change_v1[dim]
+        v2[dim] += change_v1[dim]  # conservation of momentum
+
+cdef void sub_update(FLOAT[:, ::1] XY,
+                     FLOAT[:, ::1] V,
+                     float R,
+                     int i, int count,
+                     UINT[:, ::1] Grid,
+                     float grid_spacing,
+                     omp_lock_t *locks) nogil:
+    cdef:
+        FLOAT *XY1, *XY2, *V1, *V2
+        int j, dim
+        float eps = 1e-5
+        unsigned int x_grid, y_grid, grid_size
+
+    # SUBPROBLEM 4: Add locking
+    acquire(&(locks[i]))
+    XY1 = &(XY[i, 0])
+    V1 = &(V[i, 0])
+    #############################################################
+    # IMPORTANT: do not collide two balls twice.
+    ############################################################
+    # SUBPROBLEM 2: use the grid values to reduce the number of other
+    # objects to check for collisions.
+    x_grid = <unsigned int>(XY[i,0] / grid_spacing)
+    y_grid = <unsigned int>(XY[i,1] / grid_spacing)
+    grid_size = <unsigned int>((1.0 / grid_spacing) + 1)
+
+    # Check all points in grid that are 2 away from current ball
+    for xIdx in range(x_grid+1, x_grid+3):
+        for yIdx in range(y_grid+1, y_grid+4+x_grid-j):
+            if xIdx < grid_size and yIdx < grid_size and Grid[xIdx,yIdx] < count:
+                acquire(&(locks[Grid[xIdx,yIdx]]))
+                XY2 = &(XY[Grid[xIdx, yIdx], 0])
+                V2 = &(V[Grid[yIdx, xIdx], 0])
+                if overlapping(XY1, XY2, R):
+                    # SUBPROBLEM 4: Add locking
+                    if not moving_apart(XY1, V1, XY2, V2):
+                        collide(XY1, V1, XY2, V2)
+
+                    # give a slight impulse to help separate them
+                    for dim in range(2):
+                        V2[dim] += eps * (XY2[dim] - XY1[dim])
+                release(&(locks[Grid[xIdx,yIdx]]))
```

Comment:
Great work on code for P5. I implemented locking using min max since I was facing deadlock once in a while using the approach you have used, but since you did most results I think it should be fine.